File: swift/grep_mt_tuned.swift
// swiftgrep_std_mt_tuned - multithreaded + the memory pillar applied. A fixed
// pool of N worker threads each pulls files from a shared atomic index and
// REUSES one growable read buffer (+ one lowercase buffer, one output buffer)
// across every file it processes. Reads a 64 KB prefix first, NUL-checks it, and
// reads the rest only if the prefix passed (so a huge binary is never faulted
// in then discarded). Swift Array<UInt8> is a mutable value type, so a worker
// owning its own [UInt8] and only growing it gives genuine reuse under ARC/CoW
// (no other reference => no copy on mutation). Mirrors d/grep_mt_tuned.d.
#if canImport(Glibc)
import Glibc
#endif

let PREFIX = 65536

var pat: [UInt8] = []
var lpat: [UInt8] = []
var ci = false
var recursive = false
var multi = false
var matched = Int32(0)

var files: [String] = []
var nextIdx = Int(0)
var idxLock = pthread_mutex_t()
var outLock = pthread_mutex_t()

@inline(__always)
func asciiLower(_ dst: inout [UInt8], _ src: UnsafePointer<UInt8>, _ n: Int) {
    dst.withUnsafeMutableBufferPointer { d in
        for i in 0..<n {
            let b = src[i]
            d[i] = (b >= 0x41 && b <= 0x5A) ? b &+ 32 : b
        }
    }
}

@inline(__always)
func byteIndex(_ hay: UnsafePointer<UInt8>, _ len: Int,
               _ needle: UnsafePointer<UInt8>, _ nlen: Int, _ from: Int) -> Int {
    if nlen == 0 { return from }
    if from + nlen > len { return -1 }
    guard let h = memmem(hay + from, len - from, needle, nlen) else { return -1 }
    return UnsafePointer(h.assumingMemoryBound(to: UInt8.self)) - hay
}

@inline(__always)
func lineStart(_ data: UnsafePointer<UInt8>, _ m: Int) -> Int {
    var i = m - 1
    while i >= 0 { if data[i] == 0x0A { return i + 1 }; i -= 1 }
    return 0
}

@inline(__always)
func lineEnd(_ data: UnsafePointer<UInt8>, _ m: Int, _ len: Int) -> Int {
    var i = m
    while i < len { if data[i] == 0x0A { return i }; i += 1 }
    return len
}

// Per-thread reusable scratch: each worker owns these (real reuse across every
// file the thread processes). class so it can be passed by reference into the
// worker without value-copying the backing storage.
final class Scratch {
    var data = [UInt8]()  // reused read buffer
    var low = [UInt8]()   // reused lowercase buffer
    var ob = [UInt8]()    // reused output buffer
}

func searchFile(_ s: Scratch, _ path: String) {
    let fd = open(path, O_RDONLY)
    if fd < 0 { return }
    var st = stat()
    if fstat(fd, &st) != 0 { close(fd); return }
    var len = st.st_size > 0 ? Int(st.st_size) : 0
    if s.data.count < len { s.data = [UInt8](repeating: 0, count: len) }

    // Read only the 64 KB prefix first.
    var peek = len < PREFIX ? len : PREFIX
    var off = 0
    while off < peek {
        let got = s.data.withUnsafeMutableBufferPointer { read(fd, $0.baseAddress! + off, peek - off) }
        if got <= 0 { break }
        off += got
    }
    if off < peek { peek = off; len = off }

    // NUL-check the prefix; only read the rest if it passes.
    if peek > 0 {
        let bin = s.data.withUnsafeBufferPointer { memchr($0.baseAddress!, 0, peek) != nil }
        if bin { close(fd); return }   // binary: skip, rest unread
    }

    // Read the remainder.
    while off < len {
        let got = s.data.withUnsafeMutableBufferPointer { read(fd, $0.baseAddress! + off, len - off) }
        if got <= 0 { break }
        off += got
    }
    close(fd)
    if off < len { len = off }
    if len == 0 { return }

    if ci {
        if s.low.count < len { s.low = [UInt8](repeating: 0, count: len) }
        s.data.withUnsafeBufferPointer { asciiLower(&s.low, $0.baseAddress!, len) }
    }

    let pathBytes = Array(path.utf8)
    s.ob.removeAll(keepingCapacity: true)

    s.data.withUnsafeBufferPointer { db in
        let dptr = db.baseAddress!
        let hayArr = ci ? s.low : s.data
        hayArr.withUnsafeBufferPointer { hb in
            let hay = hb.baseAddress!
            let needle = ci ? lpat : pat
            needle.withUnsafeBufferPointer { nb in
                let nptr = nb.baseAddress!
                let nlen = needle.count
                var pos = 0
                while pos < len {
                    let m = byteIndex(hay, len, nptr, nlen, pos)
                    if m < 0 { break }
                    let ls = lineStart(dptr, m)
                    let le = lineEnd(dptr, m, len)
                    matched = 1
                    if multi { s.ob.append(contentsOf: pathBytes); s.ob.append(0x3A) }
                    if le > ls { s.ob.append(contentsOf: UnsafeBufferPointer(start: dptr + ls, count: le - ls)) }
                    s.ob.append(0x0A)
                    pos = le + 1
                }
            }
        }
    }

    if !s.ob.isEmpty {
        pthread_mutex_lock(&outLock)
        s.ob.withUnsafeBufferPointer { _ = write(1, $0.baseAddress, $0.count) }
        pthread_mutex_unlock(&outLock)
    }
}

func workerBody() {
    let s = Scratch()
    while true {
        pthread_mutex_lock(&idxLock)
        let i = nextIdx
        nextIdx += 1
        pthread_mutex_unlock(&idxLock)
        if i >= files.count { break }
        searchFile(s, files[i])
    }
}

// pthread entrypoint trampoline (C ABI). No per-thread argument needed; all
// state is global, the work queue is the shared atomic index.
func threadTrampoline(_ arg: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    workerBody()
    return nil
}

func walkDir(_ dir: String) {
    guard let pd = opendir(dir) else { return }
    while let ent = readdir(pd) {
        var nameBuf = ent.pointee.d_name
        let name = withUnsafePointer(to: &nameBuf) {
            $0.withMemoryRebound(to: CChar.self, capacity: 256) { String(cString: $0) }
        }
        if name == "." || name == ".." { continue }
        collect(dir + "/" + name)
    }
    closedir(pd)
}

func collect(_ path: String) {
    var st = stat()
    if lstat(path, &st) != 0 { return }
    let mode = st.st_mode & S_IFMT
    if mode == S_IFLNK { return }
    if mode == S_IFDIR {
        if recursive { walkDir(path) }
    } else if mode == S_IFREG {
        files.append(path)
    }
}

func usage() {
    let b = Array("usage: swiftgrep_std_mt_tuned [-r] [-i] PATTERN PATH...\n".utf8)
    b.withUnsafeBufferPointer { _ = write(2, $0.baseAddress, $0.count) }
    exit(2)
}

func numThreads() -> Int {
    var n = Int(sysconf(Int32(_SC_NPROCESSORS_ONLN)))
    if n < 1 { n = 1 }
    if n > 16 { n = 16 }
    return n
}

func run() {
    var patStr = ""
    var patSet = false
    var paths: [String] = []
    var noMore = false
    for a in CommandLine.arguments[1...] {
        let ab = Array(a.utf8)
        if !noMore && ab.count >= 2 && ab[0] == 0x2D {
            if a == "--" { noMore = true; continue }
            for c in ab[1...] {
                if c == 0x69 { ci = true }
                else if c == 0x72 { recursive = true }
                else { usage() }
            }
        } else if !patSet { patStr = a; patSet = true }
        else { paths.append(a) }
    }
    if !patSet || paths.isEmpty { usage() }

    pat = Array(patStr.utf8)
    lpat = [UInt8](repeating: 0, count: pat.count)
    pat.withUnsafeBufferPointer { asciiLower(&lpat, $0.baseAddress!, pat.count) }
    multi = recursive || paths.count > 1

    for p in paths {
        var st = stat()
        if lstat(p, &st) != 0 { continue }
        let mode = st.st_mode & S_IFMT
        if mode == S_IFDIR {
            if recursive { walkDir(p) }
        } else if mode == S_IFREG {
            files.append(p)
        } else if mode == S_IFLNK {
            if stat(p, &st) == 0 {
                let m2 = st.st_mode & S_IFMT
                if m2 == S_IFREG { files.append(p) }
                else if m2 == S_IFDIR && recursive { walkDir(p) }
            }
        }
    }

    pthread_mutex_init(&idxLock, nil)
    pthread_mutex_init(&outLock, nil)

    if !files.isEmpty {
        var nt = numThreads()
        if nt > files.count { nt = files.count }
        if nt < 1 { nt = 1 }
        var tids = [pthread_t](repeating: pthread_t(), count: nt)
        for t in 0..<nt {
            pthread_create(&tids[t], nil, threadTrampoline, nil)
        }
        for t in 0..<nt {
            pthread_join(tids[t], nil)
        }
    }

    exit(matched != 0 ? 0 : 1)
}

run()

File: nim/grep_mt_tuned.nim
# nimgrep_std_mt_tuned - multithreaded + the memory pillar. Each worker reuses
# ONE growable read buffer (and one lowercase buffer, one output buffer) across
# every file it processes. Reads a 64 KB prefix first, NUL-checks it, and only
# reads the rest of the file if the prefix passed. Output serialized with a Lock
# around whole-buffer writes. Mirrors c/grep_std_mt.c and d/grep_mt_tuned.d.
#   nim c -d:release --threads:on -o:bin/nimgrep_std_mt_tuned nim/grep_mt_tuned.nim
import std/[os, posix, locks, atomics, cpuinfo]

const Prefix = 65536

var
  pat {.global.}: seq[byte]
  lpat {.global.}: seq[byte]
  ci {.global.} = false
  recursive {.global.} = false
  multi {.global.} = false
  files {.global.}: seq[string]
  nextIdx {.global.}: Atomic[int]
  matched {.global.}: Atomic[int]
  outLock {.global.}: Lock

template lowerByte(b: byte): byte =
  (if b >= byte('A') and b <= byte('Z'): byte(b + 32) else: b)

proc byteIndex(hay: openArray[byte], len: int, needle: openArray[byte], frm: int): int =
  let nlen = needle.len
  if nlen == 0: return frm
  if frm + nlen > len: return -1
  let first = needle[0]
  let last = len - nlen
  var i = frm
  while i <= last:
    if hay[i] == first:
      var j = 1
      while j < nlen and hay[i + j] == needle[j]: inc j
      if j == nlen: return i
    inc i
  -1

# Per-thread reusable scratch: each worker owns its buffers (real reuse across
# every file the thread processes).
type Scratch = object
  data: seq[byte]   # reused read buffer
  low: seq[byte]    # reused lowercase buffer
  ob: string        # reused output buffer

proc searchFile(s: var Scratch, path: string) =
  let fd = open(path.cstring, O_RDONLY)
  if fd < 0: return
  var st: Stat
  if fstat(fd, st) != 0: (discard close(fd); return)
  var len = if st.st_size > 0: int(st.st_size) else: 0
  if len <= 0: (discard close(fd); return)
  if s.data.len < len: s.data.setLen(len)

  # read only the 64 KB prefix first
  var peek = if len < Prefix: len else: Prefix
  var off = 0
  while off < peek:
    let got = read(fd, addr s.data[off], peek - off)
    if got <= 0: break
    off += got
  if off < peek: (peek = off; len = off)

  # NUL-check the prefix; only read the rest if it passes
  for i in 0 ..< peek:
    if s.data[i] == 0'u8: (discard close(fd); return)   # binary

  # read the remainder
  while off < len:
    let got = read(fd, addr s.data[off], len - off)
    if got <= 0: break
    off += got
  discard close(fd)
  if off < len: len = off
  if len <= 0: return

  let needle = if ci: lpat else: pat
  if ci:
    if s.low.len < len: s.low.setLen(len)
    for i in 0 ..< len: s.low[i] = lowerByte(s.data[i])

  let pathBytes = cast[seq[byte]](path)
  var pos = 0
  while pos < len:
    let m =
      if ci: byteIndex(s.low, len, needle, pos)
      else:  byteIndex(s.data, len, needle, pos)
    if m < 0: break
    var ls = m
    while ls > 0 and s.data[ls - 1] != byte('\n'): dec ls
    var le = m
    while le < len and s.data[le] != byte('\n'): inc le
    matched.store(1, moRelaxed)
    if multi:
      for b in pathBytes: s.ob.add char(b)
      s.ob.add ':'
    for k in ls ..< le: s.ob.add char(s.data[k])
    s.ob.add '\n'
    pos = le + 1

  if s.ob.len > 0:
    acquire(outLock)
    discard writeBuffer(stdout, addr s.ob[0], s.ob.len)
    release(outLock)
    s.ob.setLen(0)

proc worker(unused: int) {.thread.} =
  {.cast(gcsafe).}:
    var s = Scratch(ob: newStringOfCap(65536))
    while true:
      let idx = nextIdx.fetchAdd(1, moRelaxed)
      if idx >= files.len: break
      searchFile(s, files[idx])

proc isSymlink(path: string): bool =
  var st: Stat
  if lstat(path.cstring, st) != 0: return false
  S_ISLNK(st.st_mode)

proc walk(dir: string) =
  for kind, p in walkDir(dir):
    case kind
    of pcFile: files.add p
    of pcDir:
      if not isSymlink(p): walk(p)
    of pcLinkToFile, pcLinkToDir: discard

proc usage() =
  stderr.write "usage: nimgrep_std_mt_tuned [-r] [-i] PATTERN PATH...\n"
  quit(2)

proc numThreads(): int =
  result = countProcessors()
  if result < 1: result = 6
  if result > 16: result = 16

proc main() =
  var
    patStr = ""
    patSet = false
    paths: seq[string]
    noMore = false
  for a in commandLineParams():
    if not noMore and a.len >= 2 and a[0] == '-':
      if a == "--": noMore = true; continue
      for c in a[1 .. ^1]:
        if c == 'i': ci = true
        elif c == 'r': recursive = true
        else: usage()
    elif not patSet:
      patStr = a; patSet = true
    else:
      paths.add a
  if not patSet or paths.len == 0: usage()

  pat = cast[seq[byte]](patStr)
  lpat = newSeq[byte](pat.len)
  for i in 0 ..< pat.len: lpat[i] = lowerByte(pat[i])
  multi = recursive or paths.len > 1

  for p in paths:
    var st: Stat
    if stat(p.cstring, st) != 0: continue
    if S_ISDIR(st.st_mode):
      if recursive: walk(p)
    elif S_ISREG(st.st_mode):
      files.add p

  nextIdx.store(0)
  matched.store(0)
  initLock(outLock)

  var nt = numThreads()
  if nt > files.len: nt = files.len
  if nt < 1: nt = 1

  if files.len > 0:
    var threads = newSeq[Thread[int]](nt)
    for i in 0 ..< nt: createThread(threads[i], worker, i)
    joinThreads(threads)

  stdout.flushFile()
  quit(if matched.load() != 0: 0 else: 1)

main()
